File: sdk/typescript/src/instrumentors/openai.ts
/**
 * OpenAI auto-instrumentation
 */

import { getTracer, type Span } from "../tracer";

let originalSyncCreate: ((...args: unknown[]) => unknown) | null = null;
let originalAsyncCreate: ((...args: unknown[]) => unknown) | null = null;

// Pricing table (per 1M tokens)
const MODEL_PRICING: Record<string, { input: number; output: number }> = {
  "gpt-4o": { input: 2.5, output: 10.0 },
  "gpt-4o-mini": { input: 0.15, output: 0.6 },
  "gpt-4-turbo": { input: 10.0, output: 30.0 },
  "gpt-4": { input: 30.0, output: 60.0 },
  "gpt-3.5-turbo": { input: 0.5, output: 1.5 },
  "o1": { input: 15.0, output: 60.0 },
  "o1-mini": { input: 3.0, output: 12.0 },
  "o3": { input: 10.0, output: 40.0 },
  "o3-mini": { input: 1.1, output: 4.4 },
};

function calculateCost(
  model: string,
  inputTokens: number,
  outputTokens: number
): number | null {
  const pricing = MODEL_PRICING[model];
  if (!pricing) return null;
  return (
    (inputTokens / 1_000_000) * pricing.input +
    (outputTokens / 1_000_000) * pricing.output
  );
}

function buildSpan(
  model: string,
  messages: unknown[]
): Span {
  const tracer = getTracer();
  const span = tracer.startSpan("openai.chat.completions.create", {
    spanType: "llm",
    attributes: {
      "gen_ai.system": "openai",
      "gen_ai.request.model": model,
      model,
      messages,
    },
  });
  span.input = { model, messages };
  return span;
}

function finishSpan(span: Span, result: unknown, model: string): void {
  if (result && typeof result === "object") {
    const obj = result as Record<string, unknown>;

    // Extract token usage
    const usage = obj.usage as Record<string, unknown> | undefined;
    if (usage) {
      const inputTokens = (usage.prompt_tokens as number) || 0;
      const outputTokens = (usage.completion_tokens as number) || 0;
      span.tokens = {
        input: inputTokens,
        output: outputTokens,
        total: (usage.total_tokens as number) || 0,
      };

      // Calculate cost
      const cost = calculateCost(model, inputTokens, outputTokens);
      if (cost !== null) {
        span.costUsd = cost;
      }
    }

    // Extract output
    const choices = obj.choices as Array<Record<string, unknown>> | undefined;
    if (choices && choices.length > 0) {
      const choice = choices[0];
      const message = choice.message as Record<string, unknown> | undefined;
      if (message) {
        span.output = {
          content: message.content,
          role: message.role,
        };
      }
    }

    span.attributes["response_id"] = obj.id;
    span.status = "ok";
  }

  span.endTime = Date.now() / 1000;
}

export function instrument(): void {
  try {
    // Dynamic import to avoid hard dependency
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const openai = require("openai");
    if (openai?.OpenAI) {
      const OpenAI = openai.OpenAI;

      // Patch the prototype so ALL instances are instrumented
      const proto = OpenAI.prototype;
      if (proto?.chat?.completions?.create) {
        originalSyncCreate = proto.chat.completions.create;
        proto.chat.completions.create = function (
          this: unknown,
          ...args: unknown[]
        ) {
          const options = args[0] as Record<string, unknown> | undefined;
          const model = (options?.model as string) || "unknown";
          const messages = (options?.messages as unknown[]) || [];

          const span = buildSpan(model, messages);

          try {
            const result = originalSyncCreate!.apply(this, args);
            finishSpan(span, result, model);
            return result;
          } catch (error) {
            span.status = "error";
            span.attributes["error.message"] =
              error instanceof Error ? error.message : String(error);
            span.endTime = Date.now() / 1000;
            throw error;
          }
        };
      }
    }
  } catch {
    // OpenAI not available, skip instrumentation
  }
}

export function uninstrument(): void {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const openai = require("openai");
    if (openai?.OpenAI && originalSyncCreate) {
      openai.OpenAI.prototype.chat.completions.create = originalSyncCreate;
    }
  } catch {
    // Ignore
  }
  originalSyncCreate = null;
  originalAsyncCreate = null;
}

File: sdk/python/src/llm_observatory/instrumentors/openai.py
"""OpenAI auto-instrumentation."""

from __future__ import annotations

from typing import Any

from llm_observatory.pricing import calculate_cost
from llm_observatory.tracer import Span, SpanStatus, get_tracer


class OpenAIInstrumentor:
    """Instruments OpenAI API calls to capture traces."""

    def __init__(self) -> None:
        self._original_sync_create: Any | None = None
        self._original_async_create: Any | None = None

    def instrument(self) -> None:
        """Patch OpenAI to capture traces."""
        try:
            import openai

            # Patch sync completions
            self._original_sync_create = (
                openai.resources.chat.completions.Completions.create
            )
            openai.resources.chat.completions.Completions.create = self._patched_sync_create

            # Patch async completions
            self._original_async_create = (
                openai.resources.chat.completions.AsyncCompletions.create
            )
            openai.resources.chat.completions.AsyncCompletions.create = (
                self._patched_async_create
            )
        except ImportError:
            pass

    def _build_span(
        self, model: str, messages: list[dict[str, Any]]
    ) -> tuple[Span, Any]:
        """Build a span for an OpenAI call."""
        tracer = get_tracer()
        span = tracer.start_span(
            name="openai.chat.completions.create",
            span_type="llm",
        )
        span.set_input({"model": model, "messages": messages})
        span.set_attribute("gen_ai.system", "openai")
        span.set_attribute("gen_ai.request.model", model)
        return span, tracer

    def _finish_span(self, span: Span, result: Any, model: str) -> None:
        """Extract data from result and finish span."""
        try:
            # Extract token usage
            if hasattr(result, "usage") and result.usage:
                input_tokens = getattr(result.usage, "prompt_tokens", 0) or 0
                output_tokens = getattr(result.usage, "completion_tokens", 0) or 0
                span.set_token_usage(input_tokens, output_tokens)

                # Calculate cost
                cost = calculate_cost(model, input_tokens, output_tokens)
                if cost is not None:
                    span.set_cost(cost)

            # Extract output
            if hasattr(result, "choices") and result.choices:
                choice = result.choices[0]
                if hasattr(choice, "message"):
                    message = choice.message
                    output = {}
                    if hasattr(message, "content"):
                        output["content"] = message.content
                    if hasattr(message, "role"):
                        output["role"] = message.role
                    span.set_output(output)

            span.set_attribute("response_id", getattr(result, "id", None))
            span.set_status(SpanStatus.OK)

        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))

        finally:
            span.end()

    def _patched_sync_create(self, *args: Any, **kwargs: Any) -> Any:
        """Patched sync create method."""
        model = kwargs.get("model", "unknown")
        messages = kwargs.get("messages", [])

        span, _ = self._build_span(model, messages)

        try:
            result = self._original_sync_create(*args, **kwargs)
            self._finish_span(span, result, model)
            return result
        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))
            span.end()
            raise

    async def _patched_async_create(self, *args: Any, **kwargs: Any) -> Any:
        """Patched async create method."""
        model = kwargs.get("model", "unknown")
        messages = kwargs.get("messages", [])

        span, _ = self._build_span(model, messages)

        try:
            result = await self._original_async_create(*args, **kwargs)
            self._finish_span(span, result, model)
            return result
        except Exception as e:
            span.set_status(SpanStatus.ERROR)
            span.set_attribute("error.message", str(e))
            span.end()
            raise

    def uninstrument(self) -> None:
        """Remove instrumentation."""
        try:
            import openai

            if self._original_sync_create is not None:
                openai.resources.chat.completions.Completions.create = (
                    self._original_sync_create
                )
            if self._original_async_create is not None:
                openai.resources.chat.completions.AsyncCompletions.create = (
                    self._original_async_create
                )
        except ImportError:
            pass
